frontend_tests/node_tests/recent_senders.js

```diff
@@ -113,4 +115,46 @@ run_test('process_message_for_senders', () => {
         true);
 
     assert.equal(rs.compare_by_recency({}, {}, next_id += 1, ''), 0);
+
+    // new message in topic2
+    const message9 = {
+        stream_id: stream3,
+        id: next_id += 1,
+        topic: topic2,
+        sender_id: sender3,
+    };
+
+    set_global('message_util', {
+        get_messages_in_topic: (topic) => {
+            // message7's topic was changed by user
+            message7[topic] = topic3;
+            if (topic === topic2) {
+                return [message9];
+            }
+            return [message7, message8];
+        },
+    });
+
+    rs.process_topic_edit(stream3, topic2, topic3);
+    assert.equal(
+        rs.compare_by_recency({user_id: sender2}, {user_id: sender3}, stream3, topic3) > 0,
+        true);
+
+    // stream of topic3 was changed to stream4.
+    message7.stream_id = stream4;
+    message8.stream_id = stream4;
+    rs.process_topic_edit(stream3, topic3, topic3, stream4);
+    assert.equal(
+        rs.compare_by_recency({user_id: sender2}, {user_id: sender3}, stream4, topic3) > 0,
+        true);
+
+    // stream of topic3 was changed to stream5 and topic was changed to topic4.
+    message7.stream_id = stream4;
+    message8.stream_id = stream4;
+    message7.topic = topic4;
+    message8.topic = topic4;
+    rs.process_topic_edit(stream3, topic3, topic4, stream4);
+    assert.equal(
+        rs.compare_by_recency({user_id: sender2}, {user_id: sender3}, stream4, topic3) > 0,
+        true);
```

Comment:
We need these tests to verify more than they do.  Right now, if you one deletes key lines from the implementation (like deleting the data from the old topic), the tests still pass, which makes me think we haven't set these up correctly.

Reply:
Sorry for the bad work on tests. I updated tests to test what we really need from recent_senders to do for us in perspective of Recent Topics UI. Also there is a new function in recent senders `get_topic_recent_senders` which is utilized by Recent Topics.